controllers/harborconfiguration_controller.go

```diff
@@ -0,0 +1,421 @@
+		return ctrl.Result{}, err
+	}
+
+	_, err = r.replicationRuleReconciliation(ctx, harborConfiguration, *registry, client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, err
+}
+
+func getHarborSecret(ctx context.Context, clientSet *kubernetes.Clientset, harborcluster *harborOperator.HarborCluster) (string, error) {
+	passwordSecret, err := clientSet.CoreV1().Secrets(harborcluster.Namespace).Get(ctx, harborcluster.Spec.HarborAdminPasswordRef, v1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	return string(passwordSecret.Data["secret"]), nil
+}
+
+func getHarborURL(harborcluster *harborOperator.HarborCluster) string {
+	url := fmt.Sprintf("http://%s-harbor-harbor-core.%s/api/v2.0", harborcluster.Name, harborcluster.Namespace)
+	return url
+}
+
+func getTypedKubeConfig() (*kubernetes.Clientset, error) {
+	var config *rest.Config
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		config, err = getkubeConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kubeConfig, err := kubernetes.NewForConfig(config)
+
+	return kubeConfig, err
+}
+
+func getDynamicKubeConfig() (dynamic.Interface, error) {
+	var config *rest.Config
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		config, err = getkubeConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		fmt.Printf("error creating dynamic client: %v\n", err)
+		os.Exit(1)
+	}
+
+	return dynamicClient, err
+}
+
+func getkubeConfig() (*rest.Config, error) {
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		fmt.Printf("error getting user home dir: %v\n", err)
+		os.Exit(1)
+	}
+	kubeConfigPath := filepath.Join(userHomeDir, ".kube", "config")
+	fmt.Printf("Using kubeconfig: %s\n", kubeConfigPath)
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		fmt.Printf("error getting Kubernetes config: %v\n", err)
+		os.Exit(1)
+	}
+
+	return kubeConfig, err
+}
+
+func getConcreteHarborType(ctx context.Context, crdClient dynamic.ResourceInterface, harborConfiguration harborconfigurationv1alpha1.HarborConfiguration, harborTarget harborOperator.HarborCluster) (harborOperator.HarborCluster, error) {
+	harborUnstructured, err := crdClient.Get(ctx, harborConfiguration.Spec.HarborTarget.Name, v1.GetOptions{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "HarborCluster",
+			APIVersion: "v1alpha3",
+		},
+	})
+	if err != nil {
+		return harborTarget, err
+	}
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(harborUnstructured.UnstructuredContent(), &harborTarget)
+	if err != nil {
+		return harborTarget, err
+	}
+	return harborTarget, err
+}
+
+func deleteAll(ctx context.Context, harborConfiguration harborconfigurationv1alpha1.HarborConfiguration, client *apiv2.RESTClient) (ctrl.Result, error) {
+	err := client.DeleteRegistryByID(ctx, harborConfiguration.Status.RegistryId)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
If it errors out on `client.DeleteRegistryByID`, operator will never clean up _Project_ and _ReplicationPolicy_. It might be worth it to log the error but continue with all cleanups regardless. You could even collect the errors and bundle them to return together.